File: tests/rosetta/transpiler/php/dynamic-variable-names.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function parseIntStr($s, $base = 10) {
    return intval($s, intval($base));
}
$__start_mem = memory_get_usage();
$__start = _now();
  function mochi_parseIntStr($str) {
  $i = 0;
  $neg = false;
  if (strlen($str) > 0 && substr($str, 0, 1 - 0) == '-') {
  $neg = true;
  $i = 1;
}
  $n = 0;
  $digits = ['0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4, '5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9];
  while ($i < strlen($str)) {
  $n = $n * 10 + $digits[substr($str, $i, $i + 1 - $i)];
  $i = $i + 1;
};
  if ($neg) {
  $n = -$n;
}
  return $n;
};
  function main() {
  $n = 0;
  while ($n < 1 || $n > 5) {
  echo rtrim('How many integer variables do you want to create (max 5) : '), PHP_EOL;
  $line = trim(fgets(STDIN));
  if (strlen($line) > 0) {
  $n = parseIntStr($line, 10);
}
};
  $vars = [];
  echo rtrim('OK, enter the variable names and their values, below
'), PHP_EOL;
  $i = 1;
  while ($i <= $n) {
  echo rtrim('
  Variable ' . _str($i) . '
'), PHP_EOL;
  echo rtrim('    Name  : '), PHP_EOL;
  $name = trim(fgets(STDIN));
  if (array_key_exists($name, $vars)) {
  echo rtrim('  Sorry, you\'ve already created a variable of that name, try again'), PHP_EOL;
  continue;
}
  $value = 0;
  while (true) {
  echo rtrim('    Value : '), PHP_EOL;
  $valstr = trim(fgets(STDIN));
  if (strlen($valstr) == 0) {
  echo rtrim('  Not a valid integer, try again'), PHP_EOL;
  continue;
}
  $ok = true;
  $j = 0;
  $neg = false;
  if (substr($valstr, 0, 1 - 0) == '-') {
  $neg = true;
  $j = 1;
}
  while ($j < strlen($valstr)) {
  $ch = substr($valstr, $j, $j + 1 - $j);
  if ($ch < '0' || $ch > '9') {
  $ok = false;
  break;
}
  $j = $j + 1;
};
  if (!$ok) {
  echo rtrim('  Not a valid integer, try again'), PHP_EOL;
  continue;
}
  $value = parseIntStr($valstr, 10);
  break;
};
  $vars[$name] = $value;
  $i = $i + 1;
};
  echo rtrim('
Enter q to quit'), PHP_EOL;
  while (true) {
  echo rtrim('
Which variable do you want to inspect : '), PHP_EOL;
  $name = trim(fgets(STDIN));
  if (strtolower($name) == 'q') {
  return;
}
  if (array_key_exists($name, $vars)) {
  echo rtrim('It\'s value is ' . _str($vars[$name])), PHP_EOL;
} else {
  echo rtrim('Sorry there\'s no variable of that name, try again'), PHP_EOL;
}
};
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/rosetta/transpiler/FS/dynamic-variable-names.fs
// Generated 2025-07-28 10:03 +0700

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
open System

let rec parseIntStr (str: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable str = str
    try
        let mutable i: int = 0
        let mutable neg: bool = false
        if ((String.length str) > 0) && ((str.Substring(0, 1 - 0)) = "-") then
            neg <- true
            i <- 1
        let mutable n: int = 0
        let digits: Map<string, int> = Map.ofList [("0", 0); ("1", 1); ("2", 2); ("3", 3); ("4", 4); ("5", 5); ("6", 6); ("7", 7); ("8", 8); ("9", 9)]
        while i < (String.length str) do
            n <- int ((n * 10) + (int (digits.[(str.Substring(i, (i + 1) - i))] |> unbox<int>)))
            i <- i + 1
        if neg then
            n <- -n
        __ret <- n
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let mutable n: int = 0
        while (n < 1) || (n > 5) do
            printfn "%s" "How many integer variables do you want to create (max 5) : "
            let line: string = System.Console.ReadLine()
            if (String.length line) > 0 then
                n <- parseIntStr line
        let mutable vars: Map<string, int> = Map.ofList []
        printfn "%s" "OK, enter the variable names and their values, below\n"
        let mutable i: int = 1
        try
            while i <= n do
                try
                    printfn "%s" (("\n  Variable " + (string i)) + "\n")
                    printfn "%s" "    Name  : "
                    let name: string = System.Console.ReadLine()
                    if Map.containsKey name vars then
                        printfn "%s" "  Sorry, you've already created a variable of that name, try again"
                        raise Continue
                    let mutable value: int = 0
                    try
                        while true do
                            try
                                printfn "%s" "    Value : "
                                let valstr: string = System.Console.ReadLine()
                                if (String.length valstr) = 0 then
                                    printfn "%s" "  Not a valid integer, try again"
                                    raise Continue
                                let mutable ok: bool = true
                                let mutable j: int = 0
                                let mutable neg: bool = false
                                if (valstr.Substring(0, 1 - 0)) = "-" then
                                    neg <- true
                                    j <- 1
                                try
                                    while j < (String.length valstr) do
                                        try
                                            let ch: string = valstr.Substring(j, (j + 1) - j)
                                            if (ch < "0") || (ch > "9") then
                                                ok <- false
                                                raise Break
                                            j <- j + 1
                                        with
                                        | Continue -> ()
                                        | Break -> raise Break
                                with
                                | Break -> ()
                                | Continue -> ()
                                if not ok then
                                    printfn "%s" "  Not a valid integer, try again"
                                    raise Continue
                                value <- parseIntStr valstr
                                raise Break
                            with
                            | Continue -> ()
                            | Break -> raise Break
                    with
                    | Break -> ()
                    | Continue -> ()
                    vars <- Map.add name value vars
                    i <- i + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        printfn "%s" "\nEnter q to quit"
        while true do
            printfn "%s" "\nWhich variable do you want to inspect : "
            let name: string = System.Console.ReadLine()
            if (unbox<string> (name.ToLower())) = "q" then
                __ret <- ()
                raise Return
            if Map.containsKey name vars then
                printfn "%s" ("It's value is " + (string (vars.[name] |> unbox<int>)))
            else
                printfn "%s" "Sorry there's no variable of that name, try again"
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
